crates/functional-yew/src/lib.rs

```diff
@@ -0,0 +1,507 @@
+    F: FnOnce(&T1, &T2, &T3, &T4, &T5) -> Destructor,
+    Destructor: FnOnce() + 'static,
+    T1: PartialEq + 'static,
+    T2: PartialEq + 'static,
+    T3: PartialEq + 'static,
+    T4: PartialEq + 'static,
+    T5: PartialEq + 'static,
+{
+    struct UseEffectState<T1, T2, T3, T4, T5, Destructor> {
+        o1: Rc<T1>,
+        o2: Rc<T2>,
+        o3: Rc<T3>,
+        o4: Rc<T4>,
+        o5: Rc<T5>,
+        destructor: Option<Box<Destructor>>,
+    }
+    let o1 = Rc::new(o1);
+    let o2 = Rc::new(o2);
+    let o3 = Rc::new(o3);
+    let o4 = Rc::new(o4);
+    let o5 = Rc::new(o5);
+    let o1_c = o1.clone();
+    let o2_c = o2.clone();
+    let o3_c = o3.clone();
+    let o4_c = o4.clone();
+    let o5_c = o5.clone();
+    use_hook(
+        move |state: &mut UseEffectState<T1, T2, T3, T4, T5, Destructor>, hook_update| {
+            let mut should_update = !(*state.o1 == *o1
+                && *state.o2 == *o2
+                && *state.o3 == *o3
+                && *state.o4 == *o4
+                && *state.o5 == *o5);
+
+            if should_update {
+                if let Some(de) = state.destructor.take() {
+                    de();
+                }
+                let new_destructor = callback(
+                    o1.borrow(),
+                    o2.borrow(),
+                    o3.borrow(),
+                    o4.borrow(),
+                    o5.borrow(),
+                );
+                state.o1 = o1.clone();
+                state.o2 = o2.clone();
+                state.o3 = o3.clone();
+                state.o4 = o4.clone();
+                state.o5 = o5.clone();
+                state.destructor.replace(Box::new(new_destructor));
+            } else if state.destructor.is_none() {
+                should_update = true;
+                state.destructor.replace(Box::new(callback(
+                    state.o1.borrow(),
+                    state.o2.borrow(),
+                    state.o3.borrow(),
+                    state.o4.borrow(),
+                    state.o5.borrow(),
+                )));
+            }
+            return move || {
+                if should_update {
+                    hook_update(move |_: &mut UseEffectState<T1, T2, T3, T4, T5, Destructor>| true)
+                }
+            };
+        },
+        || UseEffectState {
+            o1: o1_c,
+            o2: o2_c,
+            o3: o3_c,
+            o4: o4_c,
+            o5: o5_c,
+            destructor: None,
+        },
+    )();
+}
+
+pub fn use_hook<InternalHookState, HookRunner, R, InitialStateProvider, PretriggerChange: 'static>(
+    hook_runner: HookRunner,
+    initial_state_producer: InitialStateProvider,
+) -> R
+where
+    HookRunner: FnOnce(&mut InternalHookState, Rc<dyn Fn(PretriggerChange)>) -> R,
+    InternalHookState: 'static,
+    InitialStateProvider: FnOnce() -> InternalHookState,
+    PretriggerChange: FnOnce(&mut InternalHookState) -> bool,
+{
+    return CURRENT_HOOK.with(|current_hook_field| {
```

Comment:
I refactored your approach, hope you like it!

The key changes are:
* Wrap hooks in `Rc`
* Reuse code for first hook call and subsequent hook calls
* Remove need for drop calls by shrinking the scope where we use borrowed values
* Stop wrapping hook state in `Option`

```rust
    // Extract current hook
    let (hook, render_call) = CURRENT_HOOK.with(|hook_state_holder| {
        let hook_state_holder = hook_state_holder.try_borrow_mut();
        let mut hook_state_holder = hook_state_holder.expect("Nested hooks not supported");
        let mut hook_state = hook_state_holder
            .as_mut()
            .expect("No current hook. Hooks can only be called inside functional components");

        // Determine which hook position we're at and increment for the next hook
        let hook_pos = hook_state.counter;
        hook_state.counter += 1;

        // Initialize hook if this is the first call
        if hook_pos >= hook_state.hooks.len() {
            let initial_state = Rc::new(RefCell::new(initial_state_producer()));
            hook_state.hooks.push(initial_state);
        }

        let hook = hook_state.hooks[hook_pos].clone();
        let render_call = hook_state.render_call.clone();

        return (hook, render_call);
    });

    let trigger = {
        let hook = hook.clone();
        Box::new(move |pretrigger_change: PretriggerChange| {
            let mut hook = hook.borrow_mut();
            let hook = hook.downcast_mut::<InternalHookState>();
            let hook = hook
                .expect("Incompatible hook type. Hooks must always be called in the same order");
            // It's time to call the callback. We remember whether it wants to rerender or not.
            if pretrigger_change(hook) {
                // If a render was requested, we trigger it
                render_call();
            }
        })
    };

    let mut hook = hook.borrow_mut();
    let hook = hook.downcast_mut::<InternalHookState>();
    let mut hook =
        hook.expect("Incompatible hook type. Hooks must always be called in the same order");

    // Execute the actual hook closure we were given. Let it mutate the hook state and let
    // it create a callback that takes the mutable hook state.
    hook_runner(&mut hook, trigger)
```

Reply:
I love it! :)
I'll get the small things out of the way first and then integrate your refactor.

Reply:
It's way more elegant. Integrated! :)